shadows/framework/src/main/java/org/robolectric/shadows/ShadowCryptoObject.java

```diff
@@ -0,0 +1,65 @@
+package org.robolectric.shadows;
+
+import static android.os.Build.VERSION_CODES.P;
+import static android.os.Build.VERSION_CODES.R;
+import static android.os.Build.VERSION_CODES.TIRAMISU;
+
+import android.hardware.biometrics.CryptoObject;
+import android.security.identity.IdentityCredential;
+import android.security.identity.PresentationSession;
+import java.security.Signature;
+import javax.crypto.Cipher;
+import javax.crypto.Mac;
+import org.robolectric.annotation.Implementation;
+import org.robolectric.annotation.Implements;
+
+@SuppressWarnings("UnusedDeclaration")
+@Implements(value = CryptoObject.class, isInAndroidSdk = false, minSdk = P)
+public class ShadowCryptoObject {
+
+  @Implementation(minSdk = P)
+  public void __constructor__(Signature signature) {}
+
+  @Implementation(minSdk = P)
+  public void __constructor__(Cipher cipher) {}
+
+  @Implementation(minSdk = P)
+  public void __constructor__(Mac mac) {}
+
+  @Deprecated
+  @Implementation(minSdk = R)
+  public void __constructor__(IdentityCredential identityCredential) {}
+
+  @Implementation(minSdk = TIRAMISU)
+  public void __constructor__(PresentationSession presentationSession) {}
+
+  @Implementation(minSdk = P)
+  public Signature getSignature() {
```

Comment:
I want to know what the result values of these new shadowed methods before shadowing them? I supose some of them might work without shadowing.

Reply:
There is no empty constructor in the real class so I can't really let the shadow inherit CryptoObject without having to inject garbage data (which is then stored in mObject). All real methods (basically getters) in CryptoObject use the internal mCrypto object that I want to avoid. So I don't think it's possible. 

Paradoxically all getters are nullable but all constructor parameters (that initialize mObject) are @NonNull annotated.

In short: I think this is the proper solution, let me know otherwise. Now all tests are passing by the way.

This is the real class:

```

/**
 * A wrapper class for the crypto objects supported by BiometricPrompt and FingerprintManager.
 * Currently the framework supports {@link Signature}, {@link Cipher}, {@link Mac},
 * {@link IdentityCredential}, and {@link PresentationSession} objects.
 * @hide
 */
public class CryptoObject {
    private final Object mCrypto;

    public CryptoObject(@NonNull Signature signature) {
        mCrypto = signature;
    }

    public CryptoObject(@NonNull Cipher cipher) {
        mCrypto = cipher;
    }

    public CryptoObject(@NonNull Mac mac) {
        mCrypto = mac;
    }

    /**
     * Create from a {@link IdentityCredential} object.
     *
     * @param credential a {@link IdentityCredential} object.
     * @deprecated Use {@link PresentationSession} instead of {@link IdentityCredential}.
     */
    @Deprecated
    public CryptoObject(@NonNull IdentityCredential credential) {
        mCrypto = credential;
    }

    public CryptoObject(@NonNull PresentationSession session) {
        mCrypto = session;
    }

    /**
     * Get {@link Signature} object.
     * @return {@link Signature} object or null if this doesn't contain one.
     */
    public Signature getSignature() {
        return mCrypto instanceof Signature ? (Signature) mCrypto : null;
    }

    /**
     * Get {@link Cipher} object.
     * @return {@link Cipher} object or null if this doesn't contain one.
     */
    public Cipher getCipher() {
        return mCrypto instanceof Cipher ? (Cipher) mCrypto : null;
    }

    /**
     * Get {@link Mac} object.
     * @return {@link Mac} object or null if this doesn't contain one.
     */
    public Mac getMac() {
        return mCrypto instanceof Mac ? (Mac) mCrypto : null;
    }

    /**
     * Get {@link IdentityCredential} object.
     * @return {@link IdentityCredential} object or null if this doesn't contain one.
     * @deprecated Use {@link PresentationSession} instead of {@link IdentityCredential}.
     */
    @Deprecated
    public IdentityCredential getIdentityCredential() {
        return mCrypto instanceof IdentityCredential ? (IdentityCredential) mCrypto : null;
    }

    /**
     * Get {@link PresentationSession} object.
     * @return {@link PresentationSession} object or null if this doesn't contain one.
     */
    public PresentationSession getPresentationSession() {
        return mCrypto instanceof PresentationSession ? (PresentationSession) mCrypto : null;
    }

    /**
     * @hide
     * @return the opId associated with this object or 0 if none
     */
    public final long getOpId() {
        if (mCrypto == null) {
            return 0;
        } else if (mCrypto instanceof IdentityCredential) {
            return ((IdentityCredential) mCrypto).getCredstoreOperationHandle();
        } else if (mCrypto instanceof PresentationSession) {
            return ((PresentationSession) mCrypto).getCredstoreOperationHandle();
        }
        return AndroidKeyStoreProvider.getKeyStoreOperationHandle(mCrypto);
    }
};
```